flexmeasures/data/services/sensors.py

```diff
@@ -44,3 +54,154 @@ def get_sensors(
+    return None if forecast_bdf is None else forecast_bdf.event_starts[-1]
+
+
+def get_staleness(
+    sensor: Sensor, staleness_search: dict, now: datetime, is_forecast: bool = False
+) -> timedelta | None:
+    """Get the staleness of the sensor.
+
+    The staleness is defined relative to the knowledge time of the most recent event, rather than to its belief time.
+    Basically, that means that we don't really care when the data arrived,
+    as long as the available data is about what we should be able to know by now.
+
+    :param sensor:              The sensor to compute the staleness for.
+    :param staleness_search:    Deserialized keyword arguments to `TimedBelief.search`.
+    :param now:                 Datetime representing now, used both to mask future beliefs,
+                                and to measures staleness against.
+    :param is_forecast:         Whether the sensor is a forecast sensor.
+    """
+
+    # Mask beliefs before now
+    staleness_search = staleness_search.copy()  # no inplace operations
+    beliefs_before = staleness_search.get("beliefs_before")
+    if beliefs_before is not None:
+        staleness_search["beliefs_before"] = min(beliefs_before, now)
+    else:
+        staleness_search["beliefs_before"] = now
+
+    if is_forecast:
+        staleness_search["event_starts_after"] = now
+        most_recent_knowledge_time = get_max_event_start_time(
+            sensor=sensor, staleness_search=staleness_search
+        )
+    else:
+        most_recent_knowledge_time = get_most_recent_knowledge_time(
+            sensor=sensor, staleness_search=staleness_search
+        )
+    if most_recent_knowledge_time is not None:
+        staleness = now - most_recent_knowledge_time
+    else:
+        staleness = None
+
+    return staleness
+
+
+def get_status_specs(sensor: Sensor) -> dict:
+    """Get status specs from a given sensor."""
+
+    # Check for explicitly defined status specs
+    status_specs = sensor.attributes.get("status_specs")
+    if status_specs is None:
+        # Default to status specs for economical sensors with daily updates
+        if sensor.knowledge_horizon_fnc == "x_days_ago_at_y_oclock":
+            status_specs = {"staleness_search": {}, "max_staleness": "P1D"}
+        else:
+            # Default to status specs indicating immediate staleness after knowledge time
+            status_specs = {"staleness_search": {}, "max_staleness": "PT0H"}
+    return status_specs
+
+
+def get_status(
+    sensor: Sensor,
+    now: datetime,
+    status_specs: dict | None = None,
+) -> dict:
+    """Get the status of the sensor"""
+    if status_specs is None:
+        status_specs = get_status_specs(sensor=sensor)
+    status_specs = StatusSchema().load(status_specs)
+    max_staleness = status_specs.pop("max_staleness")
+    is_forecast = status_specs.pop("is_forecast", False)
+    staleness_search = status_specs.pop("staleness_search")
+    staleness = get_staleness(
```

Comment:
I would add a type hint: `staleness: timedelta = get_staleness(` to aid people when they read this.